src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -4874,14 +4871,11 @@ public object VisitBinaryExpression(BinaryExpressionAst binaryExpressionAst)
                     return Expression.Call(CachedReflectionInfo.TypeOps_AsOperator, lhs.Cast(typeof(object)), rhs.Convert(typeof(Type)));
 
                 case TokenKind.DotDot:
-                    if(lhs.Type == typeof(string)){
-                        return Expression.Call(CachedReflectionInfo.CharOps_Range,
-                                               lhs.Convert(typeof(char)),
-                                               rhs.Convert(typeof(char)));
-                    }
-                    return Expression.Call(CachedReflectionInfo.IntOps_Range,
-                                           lhs.Convert(typeof(int)),
-                                           rhs.Convert(typeof(int)));
+                    // TODO: replace this with faster code
```

Comment:
The idea behind "faster" is to avoid extra runtime type checks.

The general pattern using `Expression.Dynamic` is to generate code like (for a binary operator like this):

```
if (lhs.GetType() == int && rhs.GetType() == int))
    return IntOps.Range((int)lhs, (int)rhs);
Find some other binding
```

If there is no binding that works, the binder generates a new site, perhaps to throw an error, or add conversions, or whatever. But the key thing here is there are no extra checks in the ideal scenario, if the lhs and rhs don't change types the next time you execute the same line in script, you have exactly the code you need.

The assumption is types don't change much, but you need to generate code that can handle that.

At any rate, it's probably not that important here, but I thought I'd expand on why the code can be made faster.

Reply:
Thanks for the deep explaination!

So if we cast all to `object` here we generate the code only one time? And now we delegate type checks to `AsChar()`. If we could move these checks from the function to this level (line 4875), then we could generate more specific sites and be faster, but I don't know if it is possible and reasonable.

Reply:
Definitely possible and should be faster, but it might only be worthwhile as an exercise in learning how `Expression.Dynamic` and call sites work.